cpp/src/arrow/dataset/expression.h

```diff
@@ -159,10 +162,27 @@ Expression call(std::string function, std::vector<Expression> arguments,
 ARROW_DS_EXPORT
 std::vector<FieldRef> FieldsInExpression(const Expression&);
 
+/// Represents either a concrete value or a hint that a field is valid/invalid
+struct KnownFieldValue {
```

Comment:
Instead, let's use a null scalar to indicate that a field is known to be null. In particular, this allows us to maintain division of responsibilities between known value replacement and constant folding, and minimizes special cases which need to be recognized and handled in simplified expressions. This is a significant change so I've assembled a PR summarizing it: https://github.com/westonpace/arrow/pull/5

Reply:
Merged in.